### PR TITLE
Use tmp folder from Magento installation to prevent path collisions

### LINKED ISSUE #29373: Hardcoded path causes collisions
https://github.com/magento/magento2/blob/9544fb243d5848a497d4ea7b88e08609376ac39e/app/code/Magento/Analytics/Model/ExportDataHandler.php#L25

This hardcoded path will cause issues when more than 1 Magento install is using the same server. It might be useful to make it unique per install, maybe using the domain URL or another unique value.

### Preconditions (*)

1. Magento Analytics installed
2. More than one Magento instance on the same server

### Steps to reproduce (*)

1. Found static path $subdirectoryPath = 'analytics/'; in dir magento2/app/code/Magento/Analytics/Model/ExportDataHandler.php
2. We used Magento 1st instance and made some minor code changes as per https://github.com/magento/magento2/issues/29373#issuecomment-885023346 in CollectData.php and ExportDataHandler.php respectively to print the dir /private/tmp/analytics/tmp/.
3. We used Magento 2nd instance and made same code changes in CollectData.php and ExportDataHandler.php respectively to print the dir /private/tmp/analytics/tmp/.

**Issue:** Both Magento instances are inserting the analytics data to the same path.

### Expected result (*)

1. Different Magento 2 installs will use different paths

### Actual result (*)

1. Different Magento 2 installs will use the same paths

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [X] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by alinalexandru**
### Description (*)

Use `tmp` folder relative to magento installation, instead of using the system tmp.
`SYS_TMP` is defined `\Magento\Framework\Filesystem\DirectoryList`, but  `\Magento\Framework\App\Filesystem\DirectoryList` is used in `\Magento\Analytics\Model\ExportDataHandler`, so it means we can us `TMP` constant.
In this way, every magento2 installation will not interfere with each other.

### Fixed Issues (if relevant)

1. Fixes magento/magento2#29373

### Manual testing scenarios (*)

I see 2 ways to do manual test, as cron cleans everything during execution
1. Follow the steps from https://github.com/magento/magento2/issues/29373#issuecomment-885023346
2. Comment the `finaly` block from `\Magento\Analytics\Model\ExportDataHandler`

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by alinalexandru** [comment]
@magento run all tests

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Please fix failing static tests

File: app/code/Magento/Analytics/Model/ExportDataHandler.php [hunk 1]
```diff
@@ -106,8 +106,10 @@ public function prepareExportData()
                 $this->cryptographer->encode($tmpDirectory->readFile($this->getArchiveRelativePath()))
             );
         } finally {
-            $tmpDirectory->delete($this->getTmpFilesDirRelativePath());
-            $tmpDirectory->delete($this->getArchiveRelativePath());
+            if (isset($tmpDirectory) && $tmpDirectory instanceof WriteInterface) {
```
Comment (Den4ik) [inline app/code/Magento/Analytics/Model/ExportDataHandler.php]: `$tmpDirecory` always will be `WriteInterface` and check `$tmpDirectory instanceof WriteInterface` is unnecessary

**Review by Den4ik — CHANGES_REQUESTED** [review]
Hi @alinalexandru 
Thanks for PR. Please look at my comment

**Comment by alinalexandru** [comment]
@Den4ik Fixed it

**Review by ihor-sviziev — APPROVED** [review]

**Review by Den4ik — APPROVED** [review]
✅ Looks good for me. All tests passed

**Comment by magento-engcom-team** [comment]
Hi @Den4ik, thank you for the review.
[ENGCOM-9245](https://jira.corp.magento.com/browse/ENGCOM-9245) has been created to process this Pull Request

**Comment by sidolov** [comment]
@magento convert to patch for 2.4.3

**Comment by sidolov** [comment]
@magento convert to patch for 2.4.3

**Comment by sidolov** [comment]
@magento convert to patch for 2.4.3

**Comment by m2-github-services** [comment]
@sidolov changes has been successfully converted for 2.4.3 version and pushed to the  converted-magento-magento2-2.4.3 branch. 

 You may submit a pull request using the link: https://github.com/magento/magento2/compare/2.4.3-release...magento:converted-magento-magento2-2.4.3

**Comment by m2-github-services** [comment]
@sidolov changes has been successfully converted for 2.4.3 version and pushed to the  converted-magento-magento2-2.4.3 branch. 

 You may submit a pull request using the link: https://github.com/magento/magento2/compare/2.4.3-release...magento:converted-magento-magento2-2.4.3

**Comment by slavvka** [comment]
@magento run all tests

**Comment by Den4ik** [comment]
@magento run Functional Tests EE

**Comment by engcom-Lima** [comment]
Hi @alinalexandru , @Den4ik , @ihor-sviziev Thanks for your contribution and collaboration .
Tried reproducing the issue in 2.4-develop version before testing the given PR fix code referring PR description/ comments.
Issue is not reproducible and below is the information on this:

**Pre Conditions**
1. Installed two magento instances with 2.4-develop version in my local with sample data.
2. Used PHP 7.4.29 version in my ubuntu system.

**Reproducible steps**
1. Followed the given steps in first magento instance [ issue comment](https://github.com/magento/magento2/issues/29373#issuecomment-885023346)
2. Repeated same steps in 2nd Magento instance as well.
3. Commented the `finally` block from `\Magento\Analytics\Model\ExportDataHandler` in both the instances.
4. Cross checked the given path `/tmp/analytics/tmp/` in the local system .

**Observation** - Analytics folder was not found in the given path `/tmp/analytics/tmp/`

**Kindly let me know based on what factors files will get generated in given path, So, that we can proceed further in our testing.**

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Preconditions:**

1. Should use the local system in which we should have all Magento instance relevant installed.
2. We should have 2 Magento latest versioned instances in the same system to test.


**Manual testing scenario:**

1. There are code changes done in 2 of the php files [commented here](https://github.com/magento/magento2/issues/29373#issuecomment-885023346), we should do the same code changes in both the instances of our Magento local code.
2. Download the `n98-magerun2-latest.phar` file from [here](https://files.magerun.net/).
3. Rename the file `n98-magerun2-latest.phar` to `n98-magerun2.phar`.
4. Copy the downloaded file inside both the local Magento instances' root folder.
5. Run the command `php -d upload_tmp_dir=/tmp n98-magerun2.phar sys:cron:run analytics_collect_data` from the Magento root folder of both instances one by one.

**Before:** :heavy_multiplication_x: The responded command used to return the folder path from both the instances which is a common local system TEMP folder.
image


**After:** :heavy_check_mark: The responded command returns the folder path from both the instances which is q unique var/ folder of each Magento instance only.
image


Since it is a specific case where the reporting folder has been changed, and so no additional regression test case are required for this case.
Note: I think, it is good to have the Magento doc to be updated with this information, even i couldn't find any info relevant to it as of now.

**Pull request merged**